be/src/formats/parquet/stored_column_reader.cpp

```diff
@@ -477,14 +543,43 @@ Status RequiredStoredColumnReader::read_records(size_t* num_records, ColumnConte
     return Status::OK();
 }
 
-Status RequiredStoredColumnReader::_next_page() {
+Status RequiredStoredColumnReader::_next_selected_page(size_t records_to_read, size_t* records_to_skip) {
+    *records_to_skip = 0;
     do {
         RETURN_IF_ERROR(_reader->next_page());
-        _num_values_left_in_cur_page = _reader->num_values();
-    } while (_num_values_left_in_cur_page == 0);
+        size_t num_values = _reader->num_values();
+        if (num_values == 0) {
+            RETURN_IF_ERROR(_reader->skip_page());
+            continue;
+        }
+        if (_select_page(num_values)) {
+            RETURN_IF_ERROR(_reader->load_page());
+            _num_values_left_in_cur_page = num_values;
+            _opts.context->next_row += num_values;
+            break;
+        }
+        RETURN_IF_ERROR(_reader->skip_page());
+        *records_to_skip += num_values;
+        _opts.context->next_row += num_values;
+    } while (*records_to_skip < records_to_read) ;
+
+    if (*records_to_skip > records_to_read) {
+        *records_to_skip = records_to_read;
+    }
     return Status::OK();
 }
 
+bool RequiredStoredColumnReader::_select_page(size_t num_values) {
+    auto selection = _opts.context->selection;
+    if (!selection) {
+        return true;
+    }
+
+    size_t start_row = _opts.context->next_row;
```

Comment:
Do you think we need a method `ColumnReaderContext::fetch_add(int num)` for simplification? I'm afraid sometimes we forget to advance the `ColumnReaderContext::next_row`.

Reply:
Yes, it seems simpler and clearer, I will modify it.